packages/next/client/use-intersection.tsx

```diff
@@ -68,9 +68,19 @@ export function useIntersection<T extends Element>({
     }
   }, [visible])
 
-  useEffect(() => {
-    if (rootRef) setRoot(rootRef.current)
-  }, [rootRef])
+  /**
+   * You can update the state right during rendering. React will re-run the component with updated state immediately after exiting the first render so it wouldn’t be expensive.
+   *
+   * An update during rendering is exactly what getDerivedStateFromProps has always been like conceptually
+   *
+   * https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops
+   */
+  if (rootRef) {
+    if (rootRef.current !== root) {
+      setRoot(rootRef.current)
+    }
+  }
+
   return [setRef, visible, resetVisible]
 }
```

Comment:
Do not use `useEffect` to sync props to state. React allows updating a state during the render phase (and it behaves just like `getDerivedStateFromProps`).

Before:

- The parent component `<Image />` receives an update, and `useIntersection` receives a new `rootRef`.
- The parent component `<Image />` finishes the update and render.
- The side-effect commit happens **after** the render phase, and the new `rootRef` is synced to state `root`.
- The identity of `setRef` changes (since `setRef` is wrapped inside a `useCallback` with `root` inside the dependencies array).
- Because the identity of `setRef` changes, the parent component `<Image />` has to update again, causing the extra render.

After:

- The parent component `<Image />` receives an update, and `useIntersection` receives a new `rootRef`.
- During `useIntersection` execution, the `root` state updates.
- React immediately exit the updating of `<Image />` components, and `<Image />` re-start rendering.
- And finally, the `<Image />` component only renders once.

Reply:
Is it possible to add a test case to avoid regression? Thanks!

Reply:
I am wondering if it is possible to test against the rendering counts of a component using the existing test suite. That's why I didn't bring up a test case along with the PR.